core/src/main/java/io/questdb/griffin/engine/functions/groupby/FirstStringGroupByFunction.java

```diff
@@ -0,0 +1,105 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2022 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.griffin.engine.functions.groupby;
+
+import io.questdb.cairo.ArrayColumnTypes;
+import io.questdb.cairo.ColumnType;
+import io.questdb.cairo.map.MapValue;
+import io.questdb.cairo.sql.Function;
+import io.questdb.cairo.sql.Record;
+import io.questdb.griffin.engine.functions.*;
+import io.questdb.std.str.DirectCharSink;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class FirstStringGroupByFunction extends StrFunction implements GroupByFunction, UnaryFunction {
+    protected final Function arg;
+    protected int valueIndex;
```

Comment:
stringValues could grow uncontrollably and cause OOM for large tables and keyed group by's string .
Could you replace it with native allocation and add limit(s) on max amount of memory it uses to configuration ?
For example - please have a look at RndStringMemory . 
It stores strings in one area and list of indexes to next string in another . 
For first(string) it should be straightforward because value is set once .
For last(string) value might get overwritten many times with strings of varying length . 
To avoid moving memory constantly (I mean all strings following current one) it would make sense to e.g. keep position, length in index list and think about re-using memory 'freed' because value had to be moved.

Reply:
Started a discussion on Slack, think I need some more ideas/explanations here